ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/CommandDeduplicationIT.scala

```diff
@@ -97,6 +100,16 @@ final class CommandDeduplicationIT(
         response.completion.commandId == request.commands.get.commandId,
         "The command ID of the first completion does not match the command ID of the submission",
       )
+      if (!ledger.features.commandDeduplicationFeatures.deduplicationType.isSyncOnly) {
+        val completion = assertCompletionIsDefined(optCompletion)
+        assertDeduplicationDuration(
```

Comment:
This check does not make much sense when durations are converted into offsets:
1. You're not looking at the offset that is reported in the completion. So the answer is independent of the actually reported deduplication period.
2. Imagine that the requested deduplication period is 100 years rather than 2s. There is likely no offset that was published a 100 years ago (because the ledger hasn't been deployed for so long. So what should the ledger return? At best, it can return the genesis offset that was published 5 seconds ago, and that should be considered fine.
3. For `ALREADY_EXISTS` completions, there's little point in checking that the deduplication period has not been shorted. This is because the semantics would be the same for any larger deduplication period, because they all include the conflicting completion. Rather you want to check that there indeed is a conflicting completion inside the reported deduplication period or, if the `completion_offset` metadata field is there, at the specified offset.

Reply:
AFAICS we report the deduplication period in `ALREADY_EXISTS` completion only because the `submission_id` and `completion_offset` metadata fields are optional, so that the application knows where to start scanning for the conflicting completion. In that case, a shorter deduplication period is actually better for the application because it has to scan less.

Reply:
@andreaslochbihler-da,
Re 1. Good catch, adding `assertDefined(completion.deduplicationPeriod.deduplicationOffset, ...`.
Re 2. Is there anything I can do in this case other than add the above assertion?
Re 3. We already have assertions for `completion_offset` and `submission_id`. Apart from adding the assertion from point 1, shall I get rid of the check for shortening the deduplication period? However, it's still not clear to me why it doesn't break the deduplication guarantees.

Reply:
Re 1: Just checking that `completion.deduplicationPeriod.deduplicationOffset` is defined doesn't add much here. The test case still does not test what it should test. You want to make sure that you use the submission time that corresponds to the reported offset, not to the offset that you think it should return.

Re 2+3: Let's take a step back. To the application, we want to guarantee that if the application specifies a deduplication period `p`, then the submission should not be accepted if there is already a conflicting completion in this period. This is a statement only about completions with status code `OK`; it does not constrain the reported deduplication period for rejections at all. Moreover note that the longer the reported deduplication period, the stronger this statement (because it covers a larger time range).

For a good user experience, we still want to provide a bit more. Because it might be that the submission is rejected for some other reason than deduplication, e.g., a contract key lookup inconsistency detected on the committer. Such an error is typically handled by the application resubmitting the command. But such a resubmission is pointless if it will then be rejected with `ALREADY_EXISTS`; we do check on the Daml execution engine fairly early on that the command doesn't already exist. And that's something we want to offer to the application. That's why we also report a deduplication periods for `CommandRejected` updates. And here, we'd like to also make sure that the deduplication check has scanned the whole deduplication period that was requested. Otherwise the committer might have shortened the period, not detected the problem, but when the later resubmission has remedies the problem, it will then have to use the full deduplication period and notice that there *is* a conflicting submission.

Yet, if the rejection is `ALREADY_EXISTS` due to a conflicting completion inside the period `p`, the situation is slightly different. If all such rejections had the `completion_offset` metadata field in there, we wouldn't need to report a deduplication period for such rejections at all. Because the `completion_offset` provides a witness to the application that the committer found a conflicting completion within the deduplication period. Alas, for implementation reasons, some ledger implementations do not provide the `completion_offset`. Therefore, the semantics of the reported deduplication period for `ALREADY_EXISTS` rejections is essentially the negation of the above deduplication guarantee statement, namely that there exists a conflicting completion within the reported period. Now, the shorter this reported period is, the stronger this statement (because it constrains more where this conflicting completion is on the completion stream). That's why I'd argue that it is reasonable if a ledger implementation reports a shorter deduplication period in case it has found a conflicting submission (it may still report a longer period than requested if it likes to do so).

Reply:
So all that having been said, I'd remove the check that the deduplication period is not shortened for the completions that report a duplicate command submission. Instead, you should check that there indeed is a conflicting completion within the reported period.

Reply:
For the problem with ledger begin, I don't think that this problem manifests in this test because you check the deduplication period only for submissions where there was already an earlier successful submission.

Reply:
Thanks for the thorough explanation again. As in a private message, I think I addressed most of the issues.
I'm still not sure if I'm missing sth in case of duration -> offset conversion and a successful completion.
Also, for some reason `bazel test //ledger/ledger-api-test-tool-on-canton:conformance-test` does not hit the case when there is a duplicate (duration -> offset conversion as well, obviously), so I have to investigate that.